pkg/deploy/deploy.go

```diff
@@ -80,8 +88,101 @@ func (o *deployHandler) ApplyKubernetes(kubernetes v1alpha2.Component) error {
 // Execute will deploy the listed information in the `exec` section of devfile.yaml
 // We currently do NOT support this in `odo deploy`.
 func (o *deployHandler) Execute(command v1alpha2.Command) error {
-	// TODO:
-	// * Make sure we inject the "deploy" mode label once we implement exec in `odo deploy`
-	// * Make sure you inject the "component type" label once we implement exec.
-	return errors.New("exec command is not implemented for Deploy")
+	containerComps, err := generator.GetContainers(o.devfileObj, common.DevfileOptions{FilterByName: command.Exec.Component})
+	if err != nil {
+		return err
+	}
+	if len(containerComps) != 1 {
+		return fmt.Errorf("could not find the component")
+	}
+	containerComp := containerComps[0]
+
+	// Create a Kubernetes Job and use the container image referenced by command.Exec.Component
+	// Get the component for the command with command.Exec.Component
+	completionMode := batchv1.CompletionMode("Indexed")
+	job := batchv1.Job{
+		TypeMeta: generator.GetTypeMeta(kclient.JobsKind, kclient.JobsAPIVersion),
+		ObjectMeta: metav1.ObjectMeta{
+			// TODO: Check if the name is K8s valid.
+			Name: o.componentName + "-" + o.appName + "-" + command.Id, // TODO: Is there a function to return the standard odo names?
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  containerComp.Name,
+							Image: containerComp.Image,
+							// TODO: Should we use the command and args already defined inside the container component?
+							// TODO: What if the 'sh' binary is not always available?
+							Command:         []string{"/bin/sh"},
+							Args:            getCmdline(command),
+							ImagePullPolicy: "IfNotPresent",
+						},
+					},
+					// Set the policy to `Never` so that it keeps the pod around and they can be used to debug.
+					RestartPolicy: "Never",
+				},
+			},
+			BackoffLimit:            pointer.Int32(2),
+			CompletionMode:          &completionMode,
+			TTLSecondsAfterFinished: pointer.Int32(60),
+		},
+		Status: batchv1.JobStatus{},
+	}
+
+	// Set labels and annotations
+	job.SetLabels(odolabels.GetLabels(o.componentName, o.appName, component.GetComponentRuntimeFromDevfileMetadata(o.devfileObj.Data.GetMetadata()), odolabels.ComponentDeployMode, false))
+	job.Annotations = map[string]string{}
+	odolabels.AddCommonAnnotations(job.Annotations)
+	odolabels.SetProjectType(job.Annotations, component.GetComponentTypeFromDevfileMetadata(o.devfileObj.Data.GetMetadata()))
+
+	log.Sectionf("Executing command in container (command: %s)", command.Id)
+	spinner := log.Spinnerf("Executing %q", command.Exec.CommandLine)
```

Comment:
Just in terms of UX, I think displaying the full command-line adds some unnecessary clutter to the output. It would still be useful to `klog` the command-line, IMO.
For consistency (with the `odo dev` output for example), maybe the spinner could just display the previous message `Executing command in container (command: %s)`?

Reply:
I think we've already discussed this, but to reiterate, it helps to have a spinner running while the job completes execution, and using the same message for both `Sectionf` and spinner did not appeal to me, so I would prefer if we can show the actual command running.